drivers: eth: dsa_ksz8xxx: Add support for KSZ8463/KSZ8463F

meshium opened the pull request:
This commit adds basic support for KSZ8463/KSZ8463F chips to the dsa_ksz8xxx.c driver.

These chips have limited register compatibility with other members of the KSZ8XXX family - their registers are 16 bits wide as opposed to the 8-bit registers supported by the driver for KSZ8794 and KSZ8863. Following the general logic of the existing code, the 16-bit registers of KSZ8463 are split into 8-bit halves.

For the KSZ8463F chip, it is assumed that both ports are used in Fiber mode.

A new configuration option, CONFIG_DSA_KSZ_PORT_ISOLATING, has been added to isolate traffic between DSA slave ports.

The driver has been tested on a custom board with an STM32F7 SoC.

MaureenHelm commented:
@rlubos @jukkar @jvasanth1 can you take a look please?

@meshium please rebase

jukkar commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -259,6 +249,78 @@ static int dsa_ksz8xxx_read_static_mac_table(struct ksz8xxx_data *pdev,
+#if CONFIG_DSA_KSZ8463
+static int dsa_ksz8xxx_switch_setup(struct ksz8xxx_data *pdev)
+{
+
```
Empty line can be removed

jukkar commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -884,9 +938,9 @@ struct net_pkt *dsa_ksz8xxx_xmit_pkt(struct net_if *iface, struct net_pkt *pkt)
-	NET_DBG("TT - port: 0x%x[%p] LEN: %d 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x",
-		port_idx, iface, len, lladst.addr[0], lladst.addr[1],
-		lladst.addr[2], lladst.addr[3], lladst.addr[4], lladst.addr[5]);
+	NET_DBG("TT - port: 0x%x[%p] LEN: %d 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x", port_idx, iface, len,
```
Please change all `NET_DBG` calls to `LOG_DBG` in this file. The former is only meant for core net stack internal use.

jukkar commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -948,17 +1001,16 @@ static struct net_if *dsa_ksz8xxx_get_iface(struct net_if *iface,
 
 	ctx = net_if_l2_data(iface);
 	NET_DBG("TT - plen: %d pnum: %d pos: 0x%p dsa_port_idx: %d",
```
Same here

jukkar commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -1016,14 +1067,13 @@ static enum ethernet_hw_caps dsa_port_get_capabilities(const struct device *dev)
 {
 	ARG_UNUSED(dev);
 
-	return ETHERNET_DSA_SLAVE_PORT | ETHERNET_LINK_10BASE_T |
-		ETHERNET_LINK_100BASE_T;
+	return ETHERNET_DSA_SLAVE_PORT | ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
 }
 
 const struct ethernet_api dsa_eth_api_funcs = {
-	.iface_api.init		= dsa_iface_init,
-	.get_capabilities	= dsa_port_get_capabilities,
-	.send                   = dsa_tx,
+	.iface_api.init = dsa_iface_init,
+	.get_capabilities = dsa_port_get_capabilities,
+	.send = dsa_tx,
```
Basically all of these changes are cosmetic and do not belong to this commit. I am seeing similar changes all over this PR. If you want to tweak the look and feel, please do the changes in separate commit (could be in this PR) so that the actual changes that introduce the new driver can be clearly seen.

meshium replied:
@jukkar Cosmetic changes are the result of clang-format's work. Ok, I removed them.

meshium committed:
drivers: eth: dsa_ksz8xxx: Add support for KSZ8463/KSZ8463F

This commit adds basic support for KSZ8463/KSZ8463F chips to the
dsa_ksz8xxx.c driver.

These chips have limited register compatibility with other members
of the KSZ8XXX family - their registers are 16 bits wide as opposed
to the 8-bit registers supported by the driver for KSZ8794 and
KSZ8863. Following the general logic of the existing code,
the 16-bit registers of KSZ8463 are split into 8-bit halves.

For the KSZ8463F chip, it is assumed that both ports are used
in Fiber mode.

A new configuration option, CONFIG_DSA_KSZ_PORT_ISOLATING, has been
added to isolate traffic between DSA slave ports.

The driver has been tested on a custom board with an STM32F7 SoC.

Signed-off-by: Aleksandr Senin <al@meshium.net>

jukkar reviewed (dismissed):
LGTM

MaureenHelm commented:
@meshium can you fix the failed checks please?

meshium commented:
> @meshium can you fix the failed checks please?

@MaureenHelm I'm not sure that the failed checks is the result of my commit. how can I restart the tests?

meshium commented:
@MaureenHelm ping

meshium commented:
@tbursztyka @albertofloyd @galak @jvasanth1 @VenkatKotakonda Could you please review my PR?

decsny commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -137,7 +152,7 @@ static int dsa_ksz8xxx_probe(struct ksz8xxx_data *pdev)
 	/*
 	 * Wait for SPI of KSZ8794 being fully operational - up to 10 ms
```
need to update the comment? since it looks like now its waiting up to 1 ms?

meshium replied:
I returned the timeout in the code to the original driver's 10ms, as this may be a specific value for the ksz8794 chip.

meshium commented:
ping @MaureenHelm

albertofloyd commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -58,9 +61,15 @@ static void dsa_ksz8xxx_write_reg(const struct ksz8xxx_data *pdev,
 	};
 
+#if CONFIG_DSA_KSZ8463
+	buf[0] = KSZ8XXX_SPI_CMD_WR | (reg_addr & 0x7FF) >> 4;
```
Consider some macro definitions to avoid magic numbers

meshium replied:
fixed

MaureenHelm commented:
@jvasanth1 @jukkar @decsny please revisit

jukkar reviewed (dismissed):
Sorry for a long delay, this went under my radar as it was not showing up in my review list.

jukkar commented:
ping @jvasanth1

kartben commented:
@jvasanth1 please review/approve

decsny commented on dts/bindings/dsa/microchip,ksz8463.yaml:
is a new compatible really needed ?

maass-hamburg commented:
@meshium please rebase

lmajewski reviewed (approved):
Thanks for adding another KSZ switch chip.

maass-hamburg commented:
Microchip MEC platform doesn't matches this anymore (since https://github.com/zephyrproject-rtos/zephyr/pull/82597)
de-assigned jvasanth1 because of that

kartben commented:
@maass-hamburg @meshium could one of you two help with rebasing please?
